pkg/build/cmd/verifystarlark.go

```diff
@@ -0,0 +1,153 @@
+// The verification relies on linting frovided by the 'buildifier' binary which
+// must be in the PATH.
+// A slice of verification errors are returned, one for each file that failed verification.
+// If any execution of the `buildifier` command fails, this is returned separately.
+// commandFn is executed on every Starlark file to determine the command and arguments to be executed.
+// The caller is trusted and it is the callers responsibility to ensure that the resulting command is safe to execute.
+func verifyStarlark(ctx context.Context, workspace string, commandFn commandFunc) ([]error, error) {
+	var verificationErrs []error
+	var executionErr error
+
+	// All errors from filepath.WalkDir are filtered by the fs.WalkDirFunc.
+	// The anonymous function used here never returns an error.
+	// Lstat or ReadDir errors are reported as verificationErrors.
+	// If any execution of the `buildifier` command fails, it is reported as executionErr and
+	// any verification of subsequent files is skipped.
+	if err := filepath.WalkDir(workspace, func(path string, d fs.DirEntry, err error) error {
+		// Skip verification of the file or files within the directory if there is an error
+		// returned by Lstat or ReadDir.
+		if err != nil {
+			verificationErrs = append(verificationErrs, err)
+			return nil
+		}
+
+		if executionErr != nil {
+			return nil
+		}
+
+		if d.IsDir() {
+			return nil
+		}
+
+		if filepath.Ext(path) == ".star" {
+			command, args := commandFn(path)
+			// The caller is trusted.
+			//nolint:gosec
+			cmd := exec.CommandContext(ctx, command, args...)
+			cmd.Dir = workspace
+
+			_, err := cmd.Output()
+			if err == nil { // No error, early return.
+				return nil
+			}
+
+			// The error returned from cmd.Output() is never wrapped.
+			//nolint:errorlint
+			if err, ok := err.(*exec.ExitError); ok {
+				switch err.ExitCode() {
+				// Case comments are informed by the output of `buildifier --help`
+				case 1: // syntax errors in input
+					verificationErrs = append(verificationErrs, errors.New(string(err.Stderr)))
+					return nil
+				case 2: // usage errors: invoked incorrectly
+					executionErr = fmt.Errorf("command %q: %s", cmd, err.Stderr)
+					return nil
+				case 3: // unexpected runtime errors: file I/O problems or internal bugs
+					executionErr = fmt.Errorf("command %q: %s", cmd, err.Stderr)
+					return nil
+				case 4: // check mode failed (reformat is needed)
+					verificationErrs = append(verificationErrs, errors.New(string(err.Stderr)))
+					return nil
+				default:
+					executionErr = fmt.Errorf("command %q: %s", cmd, err.Stderr)
+					return nil
+				}
+			}
+
+			// Error was not an exit error from the command.
+			executionErr = fmt.Errorf("command %q: %v", cmd, err)
+			return nil
+		}
+
+		return nil
+	}); err != nil {
+		// Should never happen. See comment by the invocation of WalkDir for more information.
+		panic(fmt.Sprintf("unexpected error from filepath.WalkDir: %v", err))
```

Comment:
Couldn't you treat the `executionErr` as return value for the Walk here? Then you could leave that panic to the caller 🙂

Reply:
That's an interesting point. I always struggle to mentally model the WalkFunc error handling so it might take me a little while to review this suggestion and check how things behave.

Reply:
Basically any error that is not filepath.SkipDir will just abort the whole Walk-process and get returned.

Reply:
Ah! I think I had mistakenly believed that all errors from the `WalkFunc` are filtered through the next call in addition to the `WalkDir` func errors that are filtered through the `WalkFunc`

Reply:
Done in fbe033ae25c